src/include/homestore/btree/detail/btree_remove_impl.ipp

```diff
@@ -524,17 +477,12 @@ btree_status_t Btree< K, V >::merge_nodes(const BtreeNodePtr& parent_node, const
         }
 #endif
 
-        transact_write_nodes(new_nodes, leftmost_node, parent_node, context);
+        ret = transact_nodes(new_nodes, old_nodes, leftmost_node, parent_node, context);
     }
 
 out:
     // Do free/unlock based on success/failure in reverse order
-    if (ret == btree_status_t::success) {
-        for (auto it = old_nodes.rbegin(); it != old_nodes.rend(); ++it) {
-            BT_NODE_LOG(DEBUG, (*it).get(), "Freeing this node as part of successful merge");
-            free_node(*it, locktype_t::WRITE, context);
-        }
-    } else {
+    if (ret != btree_status_t::success) {
```

Comment:
Since we are doing transact_nodes which does everything include writing nodes and also freeing nodes, we no need to free nodes here and let the underlying btree (mem or index) to do that.